File: core/src/main/java/dev/vibris/core/source/SourceRecord.java
package dev.vibris.core.source;

import java.util.Arrays;
import java.util.Objects;

public final class SourceRecord {
    private final String uuid;
    private SourceState state = SourceState.VALIDATED;
    private int references;
    private boolean active;

    public SourceRecord(String uuid, int references) {
        this.uuid = Objects.requireNonNull(uuid, "uuid");
        if (uuid.isBlank()) throw new IllegalArgumentException("source UUID must not be blank");
        if (references < 1) throw new IllegalArgumentException("source must have an initial reference");
        this.references = references;
    }

    public synchronized String uuid() {
        return uuid;
    }

    public synchronized SourceState state() {
        return state;
    }

    public synchronized int references() {
        return references;
    }

    public synchronized boolean active() {
        return active;
    }

    public synchronized void retain() {
        requireState(SourceState.VALIDATED, SourceState.QUEUED, SourceState.ACTIVATING, SourceState.ACTIVE);
        references++;
    }

    public synchronized void queue() {
        transition(SourceState.QUEUED);
    }

    public synchronized void beginActivation() {
        transition(SourceState.ACTIVATING);
    }

    public synchronized void activated() {
        transition(SourceState.ACTIVE);
        active = true;
    }

    public synchronized void failed() {
        transition(SourceState.FAILED);
    }

    public synchronized void release() {
        if (references == 0) throw new IllegalStateException("source has no reference to release");
        references--;
        if (references != 0) return;
        if (active) {
            transition(SourceState.RELEASED_ACTIVE);
        } else {
            transition(SourceState.RECLAIMABLE);
        }
    }

    public synchronized void deactivate() {
        requireState(SourceState.ACTIVE, SourceState.RELEASED_ACTIVE);
        if (!active) throw new IllegalStateException("source is not active");
        active = false;
        if (references == 0) transition(SourceState.RECLAIMABLE);
    }

    public synchronized boolean deletionEligible() {
        return references == 0 && !active && state == SourceState.RECLAIMABLE;
    }

    public synchronized void beginDeleting() {
        if (!deletionEligible()) throw new IllegalStateException("source is not reclaimable");
        transition(SourceState.DELETING);
    }

    public synchronized void deleted() {
        transition(SourceState.DELETED);
    }

    private void requireState(SourceState... allowed) {
        if (Arrays.stream(allowed).noneMatch(candidate -> candidate == state)) {
            throw new IllegalStateException("invalid source state: " + state);
        }
    }

    private void transition(SourceState next) {
        boolean allowed = switch (state) {
            case VALIDATED -> next == SourceState.QUEUED || next == SourceState.FAILED ||
                next == SourceState.RECLAIMABLE;
            case QUEUED -> next == SourceState.ACTIVATING || next == SourceState.FAILED ||
                next == SourceState.RECLAIMABLE;
            case ACTIVATING -> next == SourceState.ACTIVE || next == SourceState.FAILED ||
                next == SourceState.RECLAIMABLE;
            case ACTIVE -> next == SourceState.RELEASED_ACTIVE || next == SourceState.RECLAIMABLE;
            case RELEASED_ACTIVE, FAILED -> next == SourceState.RECLAIMABLE;
            case RECLAIMABLE -> next == SourceState.DELETING;
            case DELETING -> next == SourceState.DELETED;
            case DELETED -> false;
        };
        if (!allowed) throw new IllegalStateException("invalid source transition: " + state + " -> " + next);
        state = next;
    }
}

File: core/src/main/kotlin/dev/vibris/core/source/SourceRecord.kt
package dev.vibris.core.source

class SourceRecord(
    private val uuidValue: String,
    private var referencesValue: Int,
) {
    private var stateValue = SourceState.VALIDATED
    private var activeValue = false

    init {
        require(uuidValue.isNotBlank()) { "source UUID must not be blank" }
        require(referencesValue >= 1) { "source must have an initial reference" }
    }

    @Synchronized
    fun uuid(): String = uuidValue

    @Synchronized
    fun state(): SourceState = stateValue

    @Synchronized
    fun references(): Int = referencesValue

    @Synchronized
    fun active(): Boolean = activeValue

    @Synchronized
    fun retain() {
        requireState(
            SourceState.VALIDATED,
            SourceState.QUEUED,
            SourceState.ACTIVATING,
            SourceState.ACTIVE,
            SourceState.RELEASED_ACTIVE,
        )
        referencesValue++
        if (stateValue == SourceState.RELEASED_ACTIVE) stateValue = SourceState.ACTIVE
    }

    @Synchronized
    fun queue() {
        transition(SourceState.QUEUED)
    }

    @Synchronized
    fun beginActivation() {
        transition(SourceState.ACTIVATING)
    }

    @Synchronized
    fun activated() {
        transition(SourceState.ACTIVE)
        activeValue = true
    }

    @Synchronized
    fun retryActivation() {
        transition(SourceState.QUEUED)
    }

    @Synchronized
    fun failed() {
        transition(SourceState.FAILED)
    }

    @Synchronized
    fun release() {
        check(referencesValue != 0) { "source has no reference to release" }
        referencesValue--
        if (referencesValue != 0) {
            return
        }
        if (activeValue) {
            transition(SourceState.RELEASED_ACTIVE)
        } else {
            transition(SourceState.RECLAIMABLE)
        }
    }

    @Synchronized
    fun deactivate() {
        requireState(SourceState.ACTIVE, SourceState.RELEASED_ACTIVE)
        check(activeValue) { "source is not active" }
        activeValue = false
        if (referencesValue == 0) {
            transition(SourceState.RECLAIMABLE)
        } else {
            transition(SourceState.QUEUED)
        }
    }

    @Synchronized
    fun deletionEligible(): Boolean =
        referencesValue == 0 && !activeValue && stateValue == SourceState.RECLAIMABLE

    @Synchronized
    fun beginDeleting() {
        check(deletionEligible()) { "source is not reclaimable" }
        transition(SourceState.DELETING)
    }

    @Synchronized
    fun deleted() {
        transition(SourceState.DELETED)
    }

    private fun requireState(vararg allowed: SourceState) {
        if (allowed.none { it == stateValue }) {
            throw IllegalStateException("invalid source state: $stateValue")
        }
    }

    private fun transition(next: SourceState) {
        val allowed = when (stateValue) {
            SourceState.VALIDATED ->
                next == SourceState.QUEUED ||
                    next == SourceState.FAILED ||
                    next == SourceState.RECLAIMABLE
            SourceState.QUEUED ->
                next == SourceState.ACTIVATING ||
                    next == SourceState.FAILED ||
                    next == SourceState.RECLAIMABLE
            SourceState.ACTIVATING ->
                next == SourceState.ACTIVE ||
                    next == SourceState.QUEUED ||
                    next == SourceState.FAILED ||
                    next == SourceState.RECLAIMABLE
            SourceState.ACTIVE ->
                next == SourceState.RELEASED_ACTIVE ||
                    next == SourceState.QUEUED ||
                    next == SourceState.RECLAIMABLE
            SourceState.RELEASED_ACTIVE, SourceState.FAILED -> next == SourceState.RECLAIMABLE
            SourceState.RECLAIMABLE -> next == SourceState.DELETING
            SourceState.DELETING -> next == SourceState.DELETED
            SourceState.DELETED -> false
        }
        check(allowed) { "invalid source transition: $stateValue -> $next" }
        stateValue = next
    }
}